### PR TITLE
WIP: fix input broadcasting in modeling

### LINKED ISSUE #9953: Modeling issues with input shape validation
### Description

These are all related issues:

- `Model.standard_broadcasting` is not documented.
- `_validate_input_shapes` does not use argument `validate_broadcasting` which I believe is the `Model.standard_broadcasting` flag and should turn off validation if True.
- `_validate_input_shapes` claims it checks if inputs are broadcastable. In reality it checks that the shapes of the inputs are the same. This prevents some valid use cases from running, e.g. multiply when input shapes are (10, 1) and (24,)

### Expected behavior

- When  `Model.standard_broadcasting=False` inputs with non-broadcastable shapes should be accepted (and handled by the model.

- This example should work
```
from astropy.modeling.models import math as astmath
t=np.array([np.linspace(0,1,10)]).T 
astmath.MultiplyUfunc()(t, np.arange(4))
```
### Actual behavior

```
astmath.MultiplyUfunc()(t, np.arange(4))
...
~/dev/astropy/astropy/modeling/core.py in _validate_input_shapes(inputs, argnames, n_models, model_set_axis, validate_broadcasting)
   3708     if input_shape is None:
   3709         raise ValueError(
-> 3710             "All inputs must have identical shapes or must be scalars.")
   3711 

```

### LINKED ISSUE #10170: Model evaluation crashes if some inputs are scalars
### Edited
I originally posted this because I thought it was specific to `fix_inputs`, but it's more fundamental than that, this is a revised description.

### Description

A model can be called with some inputs as scalars and some as arrays/lists, with the scalars being broadcasted to the same shape as the other inputs. After evaluation, the output is rebroadcast by `prepare_outputs()` but this checks the shapes of the outputs against the shapes of the inputs, and crashes if any of the first `n_outputs` inputs were scalars.

### Expected behavior
I should get identical results from, but the first works and the second crashes.
```
>>> g = Gaussian2D(1, 2, 3, 4, 5)
>>> g([2.5,2.5], 1)
>>> g(2.5, [1,1])
```

### Actual behavior
`ValueError: can only convert an array of size 1 to a Python scalar`

### Steps to Reproduce

```
>>> from astropy.modeling.models import Gaussian2D
>>> g = Gaussian2D(1, 2, 3, 4, 5)
>>> g([2.5,2.5], 1)
array([0.9159326, 0.9159326])
>>> g(2.5, [1,1])  # should be the same
Traceback (most recent call last):
...
ValueError: can only convert an array of size 1 to a Python scalar
```

### System Details

Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.0
Scipy 1.4.1

### PR DESCRIPTION
**Pull request by nden**
This PR is functionally complete but I marked it as WIP to get opinions on the approach. I will add documentation later.

This takes a slightly different approach to validating inputs in modeling. 
The previous behavior was:

- Input shapes were compared (in _validate_input_shapes) and *if not the same* an error was raised.
- If `Model.standard_broadcasting=True` parameters and inputs were broadcasted to a common shape.
- There was no way to specify that inputs should not be broadcast among themselves or with parameters.

The new approach:

- `Model.standard_broadcasting=True` means that broadcasting rules apply to inputs and parameters. `prepare_inputs` is not broadcasting the inputs before passing them to `Model.evaluate` (although it computes the broadcasted shape ) but let's the function implementation deal with this following numpy broadcasting rules.
- If `Model.standard_broadcasting=False` it is not assumed that inputs should broadcast nor that inputs and parameters should broadcast. Dealing with input and output shapes is left to the writer of the model. This is a rare use case but is necessary.
- A special treatment is the case of `n_outputs > n_inputs` with `standard_broadcasting=True`. In this case the shape fo the extra outputs is assumed to be the precomputed broadcasted shape.

Fixes #10170 
Fixes #9953

### CODE REVIEW

**Comment by nden** [comment]
@chris-simpson

**Comment by perrygreenfield** [comment]
It would be good to give an example of `Model.standard_broacasting=False` to make it clear (e.g., the model is doing some weird shape changing; what do they call that? Oh yeah, a shapeshifter :-).

That is, point to such a model that does that.

**Comment by pllim** [comment]
> shapeshifter :-)

Does it only happen during full moon? :D

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -2657,7 +2658,7 @@ def __call__(self, *args, **kw):
         if with_bbox and bbox is not None:
             # first check inputs are consistent in shape
             input_shape = _validate_input_shapes(args, (), self._n_models,
-                                                 self.model_set_axis, self.standard_broadcasting)
+                                                 self.model_set_axis)
```
Comment (perrygreenfield) [inline astropy/modeling/core.py:2661]: Will this cause problems if self.standard_broadcasting is False (well, it looks like this doesn't change past behavior, but still, is it not a problem?)

File: astropy/modeling/tests/test_input.py [hunk 2]
```diff
@@ -961,3 +961,113 @@ def test_call_keyword_mappings(model):
...
+    result = c([1, 1], 1)
+    expected = [[1., 1.], [0., 0.], 2.]
+    for r, e in zip (result, expected):
+        assert_allclose(r, e)
+
+    c = M(1)
+    result = c(1, [1, 1])
+    expected = [1., 0., [2., 2.]]
+    for r, e in zip (result, expected):
+        assert_allclose(r, e)
```
Comment (perrygreenfield) [inline astropy/modeling/tests/test_input.py:1073]: Need a test case where inputs are not shape compatible for a `standard_broadcasting=False`

**Pull request closed without merging**